### ISSUE TITLE
Ray initiator RayCOMMTester hanging issue using oneCCL backend

### ISSUE THREAD

**Issue by ctao456** [intel/llm-scaler#295]
### Context:
I'm trying to build a Langgraph studio agentic application that is composed of an orchestrator and two dependent running services vLLM and ComfyUI. I have been trying to bring up 2 services, vllm and ComfyUI, defined in a docker compose file. Both services run on Intel Arc Pro B60 GPUs (4x GPUs, ~31GB VRAM each) using the Level Zero runtime. 

* vllm-architect: Serves an LLM (openai/gpt-oss-20b) on GPU 0 via ZE_AFFINITY_MASK=0. 
Image: amr-registry.caas.intel.com/intelcloud/xpu-vllm-gil:1.0
* comfyui-flux2: Runs ComfyUI with HunyuanVideo 1.5 I2V for multi-GPU video generation on GPUs 2,3 via ZE_AFFINITY_MASK=2,3. 
Image: intel/llm-scaler-omni:0.1.0-b5
ComfyUI workflow: [video_hunyuan_video_1.5_i2v_multi_xpu.json](https://github.com/intel/llm-scaler/blob/main/omni/workflows/video_hunyuan_video_1.5_i2v_multi_xpu.json)

Source docker compose code:
```bash
vllm-architect:
    image: amr-registry.caas.intel.com/intelcloud/xpu-vllm-gil:1.0
    container_name: vllm_architect
    env_file:
      - ./langgraph/.env
    privileged: true
    ports:
      - "8088:8088"
    shm_size: 64g
    devices:
      - /dev/dri:/dev/dri
    environment:
      - HUGGING_FACE_HUB_TOKEN=${HUGGING_FACE_HUB_TOKEN:-}
      - HF_HOME=/llm/models
      # Intel XPU: Disable torch.compile() - Triton backend incompatible with SYCL
      - TORCH_COMPILE_DISABLE=1
      # Intel XPU: Required for multi-processing on XPU devices
      - VLLM_WORKER_MULTIPROC_METHOD=spawn
      # Intel XPU: Enumerate GPUs by PCI slot order for deterministic device assignment
      - ZE_ENABLE_PCI_ID_DEVICE_ORDER=1
      # Intel XPU: Restrict to GPU 0
      - ZE_AFFINITY_MASK=0
    volumes:
      - ./models:/llm/models
      - ./start-vllm-architect.sh:/llm/start-vllm-architect.sh
      - ./logs:/llm/logs
    entrypoint: [""]
    command: ["/bin/bash", "/llm/start-vllm-architect.sh"]
    healthcheck:
      test: ["CMD-SHELL", "curl -fsS http://localhost:8088/v1/models >/dev/null || exit 1"]
      # Poll every 10s so a ready service is detected within one interval.
      # start_period is a grace window for failures only — a passing check marks healthy immediately.
      interval: 10s
      timeout: 10s
      retries: 60
      start_period: 600s
    restart: unless-stopped
    networks:
      - lang_net

  comfyui-flux2:
    image: intel/llm-scaler-omni:0.1.0-b5
    container_name: comfyui_flux2_service
    env_file:
      - ./langgraph/.env
    privileged: true
    network_mode: host
    shm_size: 64g
    devices:
      - /dev/dri:/dev/dri
    environment:
      - HUGGING_FACE_HUB_TOKEN=${HUGGING_FACE_HUB_TOKEN:-}
      - HF_HOME=/llm/ComfyUI/models
      - PYTHONUNBUFFERED=1
      # Intel XPU: Enumerate GPUs by PCI slot order for deterministic device assignment
      - ZE_ENABLE_PCI_ID_DEVICE_ORDER=1
      # Intel XPU: GPU affinity — set COMFYUI_ZE_AFFINITY_MASK in .env ("1" = single, "2,3" = multi-GPU)
      - ZE_AFFINITY_MASK="2,3"
      # Use a different port than vllm-architect (which defaults to 29500) to avoid torch.distributed conflict
      # - MASTER_ADDR=127.0.0.1
      # - MASTER_PORT=29501
      - COMFYUI_MODEL_DIR=/llm/ComfyUI/models
      - WORKFLOW_DIR=/llm/ComfyUI/user/default/workflows
      - MODEL_DIR=/llm/ComfyUI/models
      # - http_proxy=http://proxy-dmz.intel.com:912
      # - https_proxy=http://proxy-dmz.intel.com:912
      # - HTTP_PROXY=http://proxy-dmz.intel.com:912
      # - HTTPS_PROXY=http://proxy-dmz.intel.com:912
      # - no_proxy=localhost,127.0.0.1,intel.com,.intel.com
      # - NO_PROXY=localhost,127.0.0.1,intel.com,.intel.com
    volumes:
      # Persistent model storage as requested
      - ./models:/llm/ComfyUI/models
      - ./start_comfyui.sh:/llm/ComfyUI/start_comfyui.sh
      # Mount workflow JSONs directly — same files used by the orchestrator
      - ./langgraph/orchestrator/api/video_hunyuan_video_1.5_i2v.json:/llm/ComfyUI/user/default/workflows/video_hunyuan_video_1.5_i2v.json:ro
      - ./langgraph/orchestrator/api/video_hunyuan_video_1.5_i2v_multi_xpu.json:/llm/ComfyUI/user/default/workflows/video_hunyuan_video_1.5_i2v_multi_xpu.json:ro
    # Ports are ignored when using network_mode: host
    command: [ "/bin/bash", "/llm/ComfyUI/start_comfyui.sh" ]
    healthcheck:
      # Verify ComfyUI API is responding on port 8188
      test: [ "CMD", "wget", "--no-verbose", "--tries=1", "--spider", "http://localhost:8188/" ]
      # Poll every 10s so a ready service is detected within one interval.
      # start_period is a grace window for failures only — a passing check marks healthy immediately.
      interval: 10s
      timeout: 10s
      retries: 120
      start_period: 1200s
    restart: unless-stopped
```

ComfyUI uses Raylight (a ComfyUI custom node) to distribute HunyuanVideo inference across multiple GPUs using Ray actors, and XFuser/Ulysses (degree=2) sequence parallelism. The xccl backend (oneCCL) is used for collective communication via dist.init_process_group("xccl", ...).

### Problem
When both services run concurrently, the RayCOMMTester (Raylight's pre-flight communication test) hangs indefinitely at the oneCCL ATL/OFI initialization stage. The same multi-GPU workflow completes successfully when ComfyUI runs standalone (without vLLM). We validated with a testing script calling ComfyUI API directly with a mapping of values to the workflow json.
[test_multi_gpu_video_gen.py](https://github.com/user-attachments/files/25617827/test_multi_gpu_video_gen.py)

#### Observed Behavior
ComfyUI logs show the hang at:
```log
(RayCOMMTester pid=5197) Running NCCL COMM pre-run
(RayCOMMTester pid=5197) 2026:02:27-21:25:25: 5197 |CCL_WARN| did not find MPI-launcher specific variables, switch to ATL/OFI, to force enable ATL/MPI set CCL_ATL_TRANSPORT=mpi
(RayCOMMTester pid=5197) 2026:02:27-21:25:25: 5197 |CCL_WARN| value of CCL_SYCL_CCL_BARRIER changed to be 1 (default:0)
(RayCOMMTester pid=5197) 2026:02:27-21:25:25: 5197 |CCL_WARN| could not get local_idx/count from environment variables, trying to get them from ATL
```
The process never progresses past `trying to get them from ATL`. No timeout occurs — it hangs forever.

### Root Cause Analysis
The [raylight_for_multi_arc.patch](https://github.com/intel/llm-scaler/blob/e09993b70793aedaf094ccbdd3afad34ea32cee4/omni/patches/raylight_for_multi_arc.patch#L96C8-L96C21) (applied in intel/llm-scaler-omni:0.1.0-b5) has a hardcoded device_id=0 in the RayCOMMTester and ray_nccl_tester functions:

RayCOMMTester.__init__ (from patch):
```python
class RayCOMMTester:
    def __init__(self, local_rank, world_size, device_id):
        device = torch.device(f"xpu:{device_id}")
        os.environ["ZE_AFFINITY_MASK"] = str(device_id)  # Overwrites container-level mask!

        dist.init_process_group(
            "xccl",
            rank=local_rank,
            world_size=world_size,
            timeout=timedelta(minutes=1),
            device_id=device
        )
```

ray_nccl_tester (from patch):
```python
def ray_nccl_tester(world_size):
    for local_rank in range(world_size):
        gpu_actors.append(
            gpu_actor.options(name=f"RayTest:{local_rank}").remote(
                local_rank=local_rank,
                world_size=world_size,
                device_id=0,  # <-- HARDCODED to 0 for ALL ranks
            )
        )
```

Does this mean:
1. The ComfyUI container starts with ZE_AFFINITY_MASK=2,3 (logical GPUs 0,1 mapped to physical GPUs 2,3).
2. ray_nccl_tester() spawns RayCOMMTester actors, all with device_id=0.
3. Each actor overwrites ZE_AFFINITY_MASK to "0", forcing all ranks onto physical GPU 0.

Based on the log, the execution of all_reduce failed to complete, and we were stuck in the below few lines in [Raylight `ray_worker.py`](https://github.com/komikndr/raylight/blob/36062f3a5c8d58174b4dad98d2841cc847358eb5/src/raylight/distributed_worker/ray_worker.py#L516-L519)
```python
# Each rank contributes rank+1
x = torch.ones(1, device=device) * (local_rank + 1)
dist.all_reduce(x, op=dist.ReduceOp.SUM)
result = x.item()
```

### Suggested Fix
In [raylight_for_multi_arc.patch](https://github.com/intel/llm-scaler/blob/omni-0.1.0-b5/omni/patches/raylight_for_multi_arc.patch), ray_nccl_tester and RayCOMMTester should respect the container-level ZE_AFFINITY_MASK instead of hardcoding device_id=0
```python
# ray_nccl_tester — use logical device index within the container's mask
for local_rank in range(world_size):
    gpu_actors.append(
        gpu_actor.options(name=f"RayTest:{local_rank}").remote(
            local_rank=local_rank,
            world_size=world_size,
            device_id=local_rank,  # Was: 0 (hardcoded)
        )
    )
```

Could you please suggest if related fix should be applied, or if there exists better resorts to support multiple container creations that each consumes one or multi-GPU cards for execution of vLLM and ComfyUI workflows? Thanks!

**Comment by Jasonzzt** [comment]
Hi @ctao456 ,

Thank you for the detailed report and the proposed fix. We have conducted an in-depth analysis and attempted to reproduce this behavior in a similar environment (4x Arc Pro B60). Here is our assessment:

1. Root Cause Analysis: Network Port Contention
Our findings indicate that the primary cause of the "hang" is network port contention on the default MASTER_PORT=29500 when running in network_mode: host.

The Error vs. The Hang: In our tests, we observed an explicit EADDRINUSE error, which occurs when a process attempts to bind to port 29500 as Rank 0 while it's already occupied (e.g., by vLLM).

Protocol Deadlock: The "hang" at the ATL layer (reported in your logs) typically happens when a Rank 1+ process successfully establishes a TCP connection to the port, but fails the oneCCL protocol handshake because the listener is a mismatched service. Since the ATL layer lacks a default handshake timeout, it hangs indefinitely.

2. Assessment of Proposed Code Changes
We have reviewed your proposed changes in the patch and agree they represent best practices for environment robustness, even though they might not be the direct trigger for the communication hang.

Removal of os.environ modification: Agreed for code cleanliness. Our testing shows that in a Ray-managed environment, the device isolation is already handled at the process level before the Actor starts. Manually overwriting ZE_AFFINITY_MASK in Python is redundant. While it didn't negatively impact our results, removing it is a cleaner approach that avoids unnecessary side effects and stays aligned with standard resource orchestration.

Changing device_id=0 to local_rank: Why it currently works: In a Ray sandbox (num_gpus=1), device_id=0 acts as a relative logical index. Ray masks the environment so each actor sees only its assigned GPU as "Device 0".

3. Recommended Resolution
The functional fix to resolve the hang is to isolate the communication ports between services. We suggest use a unique port for llm-scaler-omni:

```
export MASTER_PORT=29501
```
Could you please confirm if isolating the MASTER_PORT resolves the hang in your environment? We are happy to incorporate your suggested code improvements for better robustness regardless.
